### PR TITLE
Add shims for AQEShuffleReader [databricks]

### PR DESCRIPTION
**Pull request by andygrove**
Part of https://github.com/NVIDIA/spark-rapids/issues/9753

This PR updates GpuCustomShuffleReaderExec to match the logic in Spark, which has changed since we first implemented this.

There are three different sets of behaviors:

- Spark 3.1: No special handling for coalesced shuffle reads
- Spark 3.2+: Special handling for coalesced shuffle reads
- Spark 3.4.2+/3.5.1: Introduces CoalescedBoundary when reading HashPartitioned shuffles

### CODE REVIEW

File: sql-plugin/src/main/spark342/scala/com/nvidia/spark/rapids/shims/AQEShuffleReaderShims.scala [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+  def coalescedReadOutputPartitioning(partitionSpecs: Seq[ShufflePartitionSpec],
+      outputPartitioning: Partitioning): Partitioning = {
+    // For coalesced shuffle read, the data distribution is not changed, only the number of
+    // partitions is changed.
+    outputPartitioning match {
+      case h: HashPartitioning =>
+        val partitions = partitionSpecs.map {
+          case CoalescedPartitionSpec(start, end, _) => CoalescedBoundary(start, end)
+          // Can not happend due to isCoalescedRead
```
Comment (andygrove) [inline sql-plugin/src/main/spark342/scala/com/nvidia/spark/rapids/shims/AQEShuffleReaderShims.scala:48]: comment copied as is from Spark version of this code

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
The premerge failed due to 
```

[2024-02-29T21:09:42.364Z] [INFO] Compiling 450 Scala sources and 58 Java sources to /home/ubuntu/spark-rapids/sql-plugin/target/spark330db/classes...

[2024-02-29T21:09:48.866Z] [ERROR] /home/ubuntu/spark-rapids/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGetJsonObject.scala:19: object parsing is not a member of package util

[2024-02-29T21:09:48.867Z] [ERROR] import scala.util.parsing.combinator.RegexParsers
```
This should be fixed now as the Databricks nightly builds are passing. Kicking off the build again

**Comment by andygrove** [comment]
build

**Review by revans2 — COMMENTED** [review]
Do we have tests for this? All of our existing tests pass without any of these changes, so I am curious if there is a good way for us to verify that these changes are working.

**Comment by andygrove** [comment]
> Do we have tests for this? All of our existing tests pass without any of these changes, so I am curious if there is a good way for us to verify that these changes are working.

I don't think that we can reproduce the issue from https://issues.apache.org/jira/browse/SPARK-45592 until we implement https://github.com/NVIDIA/spark-rapids/issues/9876, but I could be wrong

**Comment by razajafri** [comment]
But #9876 is talking about replacing `TableCacheQueryStageExec` which we should never replace as I understand we shouldn't replace query stage execs?

**Comment by andygrove** [comment]
> But #9876 is talking about replacing `TableCacheQueryStageExec` which we should never replace as I understand we shouldn't replace query stage execs?

I think #9876 is really about making the plan inside `TableCacheQueryStageExec` run on the GPU. We can't do this today because Spark is looking for a specific `InMemoryTableScan` class so does not recognize our GPU version.

**Comment by razajafri** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuInMemoryTableScanExec.scala [hunk 2]
```diff
@@ -69,7 +72,10 @@ class InMemoryTableScanMeta(
    * Convert InMemoryTableScanExec to a GPU enabled version.
    */
   override def convertToGpu(): GpuExec = {
-    GpuInMemoryTableScanExec(imts.attributes, imts.predicates, imts.relation)
+    GpuInMemoryTableScanExec(
+      imts.attributes,
+      predicates.map(_.convertToGpu()),
+      imts.relation)
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuInMemoryTableScanExec.scala:78]: Just like with my previous comment. Can we please take a look at this and understand what is happening here? It looks like some new predicates are being pushed down into the in memory table scan exec. If so I would like to understand better what is happening here with it and what changed.

My biggest problem with this change is that we are not tagging the predicates before trying to convert them to the GPU. what happens if one of them is something that we cannot convert to the GPU? We should fall back to the CPU in those cases.

Also do we even support predicates in our GpuInMemoryTableScanExec?

**Pull request closed without merging**